File: kotlin/hamming/src/main/kotlin/Hamming.kt
object Hamming {
    fun compute(leftStrand: String, rightStrand: String): Int {
        require (leftStrand.length == rightStrand.length) {
            "leftStrand and rightStrand must be of equal length."
        }
        return (leftStrand zip rightStrand)
                .count { it.first != it.second }
    }

}

File: typescript/hamming/hamming.ts
export default class Hamming {

    compute(first: string, second: string) {
        return Hamming.zip(first, second)
            .filter(([f, s]) => f !== s)
            .length
    }

    private static throwOnLengthMismatch(first: string, second: string) {
        if (first.length !== second.length) {
            throw "DNA strands must be of equal length."
        }
    }

    private static zip(first: string, second: string) {
        Hamming.throwOnLengthMismatch(first, second)
        return first.split("")
            .map((c, index) => [c, second[index]])
    }
}